### PR TITLE
fix: make sure to send OnAppRegistered notification after resumption

### PR DESCRIPTION
**Pull request by shoamano83**
Fixes https://github.com/smartdevicelink/sdl_core/issues/2165

This PR is **ready** for review.

### Risk
This PR makes **no** API changes.

### Testing Plan
N/A

### Summary
This PR changes the order of resumption and sending BC.OnAppRegistered notification, so that BC.OnAppRegistered is always notified after resumption runs.

##### Breaking Changes
* None

##### Enhancements
* None

##### Bug Fixes
* If HMI issues SDL.ActivateApp right after BC.OnAppRegistered notification, the request may be processed before resumption runs. This caused HMI level to change again after ActivateApp is processed.

### Tasks Remaining:
- None.

### CLA
- [X] I have signed [the CLA](https://docs.google.com/forms/d/e/1FAIpQLSdsgJY33VByaX482zHzi-xUm49JNnmuJOyAM6uegPQ2LXYVfA/viewform)

### CODE REVIEW

**Comment by shoamano83** [comment]
We have applied this fix internally. It seems that we forgot to contribute to upstream, sorry about that.

Unit test is not available as this is a timing issue and cannot be reproduced reliably. What we did for verification is to add some `sleep()` to block the thread of `ResumeCtrlImpl::StartResumption()`, then issue SDL.ActivateApp from HMI as soon as possible.

File: src/components/application_manager/src/commands/mobile/register_app_interface_request.cc [hunk 1]
```diff
@@ -736,6 +734,9 @@ void RegisterAppInterfaceRequest::SendRegisterAppInterfaceResponseToMobile(
     resumer.StartResumptionOnlyHMILevel(application);
   }
 
+  SendOnAppRegisteredNotificationToHMI(
+      *(application.get()), resumption, need_restore_vr);
```
Comment (jacobkeeler) [inline src/components/application_manager/src/commands/mobile/register_app_interface_request.cc]: This results in the resumption commands (AddCommand, etc.) being sent before `OnAppRegistered` is sent to the HMI. This means that the HMI would need to add commands for an app that it doesn't know exists yet, which causes errors with the sample HMIs.

**Comment by shoamano83** [comment]
Hi @jacobkeeler , thank you for pointing out the resumption issue.

Let me clarify:
- RPCs for resumption (e.g. AddCommand) have to be issued after HMI level for the app is configured
- HMI requires OnAppRegistered notification before receiving RPCs for resumption

If I understand the implementation correctly, then current sequence is something like:

1. Send RegisterAppInterface response to mobile
2. Notify HMI of BC.OnAppRegistered
3. Restore default HMI level from policy table
4. Send AddCommand, etc. to HMI for resumption
5. Start resumption timer
6. Once the timer is fired and the app hasn't been activated yet, then the app's state (including HMI level) is restored

This is still an idea but probably we can modify the sequence like below? I would like to seek for a solution which will minimize the impact and required change.

1. Send RegisterAppInterface response to mobile
2. Restore default HMI level from policy table
3. Start resumption timer
4. Notify HMI of BC.OnAppRegistered
5. Send AddCommand, etc. to HMI for resumption
6. Once the timer is fired and the app hasn't been activated yet, then the app's state (including HMI level) is restored

Thanks.

**Comment by shoamano83** [comment]
Hi @jacobkeeler ,

I investigated the code some more and came up with another idea. The new approach is to delay SDL.ActivateApp getting processed until resumption is triggered.
I think the risk of this change is minimum since we don't change RAI sequence at all and it only affects SDL.ActivateApp. The only downside is that we add two public methods in `Application` class just for this feature, and the code size is increased.
I tested the change internally and confirmed that it fixes the original issue https://github.com/smartdevicelink/sdl_core/issues/2165 we observed.

Can I ask for another review please? I rebased the commits onto latest develop and I hope the changes are clear.

Thanks.

File: src/components/application_manager/include/application_manager/application.h [hunk 2]
```diff
@@ -1224,6 +1224,24 @@ class Application : public virtual InitialApplicationData,
    */
   virtual const smart_objects::SmartObject& get_user_location() const = 0;
 
+  /*
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/application.h:1227]:
```suggestion
  /**
```

File: src/components/application_manager/include/application_manager/application_impl.h [hunk 3]
```diff
@@ -505,6 +505,24 @@ class ApplicationImpl : public virtual Application,
       const smart_objects::SmartObject& user_location) OVERRIDE;
 
   const smart_objects::SmartObject& get_user_location() const OVERRIDE;
+  
+  /*
```
Comment (jacobkeeler) [inline src/components/application_manager/include/application_manager/application_impl.h:509]:
```suggestion
  /**
```

File: src/components/application_manager/test/application_impl_test.cc [hunk 4]
```diff
@@ -879,6 +881,50 @@ TEST_F(ApplicationImplTest, SetDeferredResumptionHMILevelTest) {
...
+    virtual void exitThreadMain() {}
+
+   private:
+    std::shared_ptr<ApplicationImpl> app_;
+    std::shared_ptr<TestAsyncWaiter> waiter_;
+    unsigned int seconds_;
+  };
+
+  std::shared_ptr<TestAsyncWaiter> waiter =
+      std::make_shared<TestAsyncWaiter>();
```
Comment (jacobkeeler) [inline src/components/application_manager/test/application_impl_test.cc:911]:
```suggestion
  std::shared_ptr<TestAsyncWaiter> waiter = std::make_shared<TestAsyncWaiter>();
```

File: src/components/application_manager/src/application_impl.cc [hunk 5]
```diff
@@ -1123,4 +1124,30 @@ void ApplicationImpl::SwapMobileMessageQueue(
...
+
+  sync_primitives::AutoLock auto_lock(setup_in_progress_lock_);
+  bool previous_flag = setup_in_progress_;
+  setup_in_progress_ = in_progress;
+  LOG4CXX_DEBUG(logger_,
+                "setup_in_progress_ of app " << app_id_ << " updated to "
+                                             << setup_in_progress_);
+
+  // unblock WaitForSetupDone() when the flag changes from true to false
+  if (previous_flag == true && setup_in_progress_ == false) {
```
[the remark this answers is not in the document]
Comment (shoamano83) [reply src/components/application_manager/src/application_impl.cc]: Thanks for your suggestion. I updated the code in commit ba0620857f0e15103acac523b9319704d8c1174c.

File: src/components/application_manager/src/application_impl.cc [hunk 6]
```diff
@@ -1123,4 +1124,30 @@ void ApplicationImpl::SwapMobileMessageQueue(
...
+    setup_in_progress_cond_.Broadcast();
```
[the remark this answers is not in the document]
Comment (shoamano83) [reply src/components/application_manager/src/application_impl.cc:1360]: Hmm, I think we have no issue calling `Broadcast` with the lock held. In pthread library, it is valid and has no issue. (Actually some people prefer calling `pthread_cond_broadcast()` with the lock held.)
Reference: https://stackoverflow.com/questions/1105250/call-pthread-cond-broadcast-with-mutex-held-or-not

Regarding `AssertFreeAndMarkTaken`:

the function `wait()` acquires the lock before it returns. So if we release the lock after calling Broadcast, `wait()` just waits until the lock is released, and returns. Hence "assert free" is always true.

I do not have strong opinion on this, so if you still think we should unlock first then just let me know. I'll update the code.

**Comment by iCollin** [comment]
Closed via https://github.com/smartdevicelink/sdl_core/pull/3559

Please feel free to reopen if the issue is reproduced.

**Pull request closed without merging**